### PR TITLE
Support Head Command In PPL

### PR DESCRIPTION
**Pull request by rupal-bq**
*Issue #684*

*Description of changes:*
- core: added support for head operation in a logical & physical layer
- ppl: added AstBuilder logic for head command
- updated documentation with examples for using head command

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/ast/dsl/AstDSL.java [hunk 1]
```diff
@@ -308,4 +332,5 @@ public static RareTopN rareTopN(UnresolvedPlan input, CommandType commandType,
     return new RareTopN(input, commandType, noOfResults, Arrays.asList(fields), groupList)
         .attach(input);
   }
+
```
Comment (lyndonb-bq) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/ast/dsl/AstDSL.java:335]: Nit: Extra line

**Comment by dai-chen** [comment]
Hi @rupal-bq , I just merged my PR for adding explain API. Now we can explain the physical plan for SQL/PPL query. Could you also add explain for Head operator? Explain for RareTopN operator for your reference: https://github.com/opendistro-for-elasticsearch/sql/blob/develop/core/src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/Explain.java#L131. UT is in `ExplainTest`.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  @Override
+  public boolean hasNext() {
+    while (input.hasNext() && !foundFirstFalse) {
+      ExprValue inputVal = input.next();
+      ExprValue exprValue = whileExpr.valueOf(inputVal.bindingTuples());
+
+      if (recordCount < number) {
+        if (!(!(exprValue.isNull() || exprValue.isMissing()) && (exprValue.booleanValue()))) {
```
Comment (lyndonb-bq) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java:91]: This if statement logically equivalent to:
`if (isNull || isMissing || !booleanValue)`
Would suggest simplifying it to that

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/ast/dsl/AstDSL.java (hunk 1 above)
Comment (rupal-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/ast/dsl/AstDSL.java]: removed

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java (hunk 2 above)
Comment (rupal-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java]: simplified above logic and removed while loop

**Comment by rupal-bq** [comment]
> Hi @rupal-bq , I just merged my PR for adding explain API. Now we can explain the physical plan for SQL/PPL query. Could you also add explain for Head operator? Here is explain for RareTopN operator for your reference: https://github.com/opendistro-for-elasticsearch/sql/blob/develop/core/src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/Explain.java#L131. UT is in `ExplainTest`.

Thanks, Chen. Added explain for Head operator.

**Review by lyndonb-bq — APPROVED** [review]

**Review by lyndonb-bq — APPROVED** [review]

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperatorTest.java [hunk 3]
```diff
@@ -1,3 +1,18 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (penghuo) [inline core/src/test/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperatorTest.java:2]: 2019 -> 2020

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprValue;
+import com.amazon.opendistroforelasticsearch.sql.expression.Expression;
+import com.amazon.opendistroforelasticsearch.sql.expression.LiteralExpression;
+import java.util.Collections;
+import java.util.List;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NonNull;
+
+@Getter
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java:28]: Doc

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperatorTest.java (hunk 3 above)
Comment (rupal-bq) [reply core/src/test/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperatorTest.java]: fixed

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java (hunk 4 above)
Comment (rupal-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/HeadOperator.java:35]: added

**Review by dai-chen — APPROVED** [review]
LGTM. Thanks!

**Review by penghuo — APPROVED** [review]
Thanks for the change.

**Pull request merged**